File: python/heritage/application/backtest_calibration_app.py
# ============================================================
# 文件：backtest_calibration_app.py
# 职责：回测与参数校准应用 — 在隔离的事件总线上重放历史行情，
#       比对"市场真实价格"与"模型预测价格"，优化隐含波动率 σ，
#       最终通过 ParamUpdateEvent 将校准结果发布到系统。
#
# 对应 Risk_Calibration.md §2（Backtest & Calibration Application）
#
# 参数反馈闭环（Risk_Calibration.md §2.7）：
#   BacktestCalibrationApp
#         ↓ 发布 ParamUpdateEvent（到主总线）
#   ParameterStore（参数仓库，存储带时间戳的校准结果）
#         ↓
#   实时定价引擎（查询 get_params() 获取最新 σ）
#
# 事件订阅：MarketDataEvent（在隔离的 backtest_bus 上）
# 事件发布：ParamUpdateEvent（发布到主总线 main_bus）
# ============================================================

from __future__ import annotations
from dataclasses import dataclass
from datetime import datetime

from events.event_bus import EventBus
from events.events import MarketDataEvent, ParamUpdateEvent
from domain.instrument import Option
from domain.pricing_engine import BlackScholesPricingEngine
from domain.calibration_engine import CalibrationEngine


@dataclass
class _RawObs:
    """原始观测：保存每个 tick 的（标的价格、期权合约、市场价格），
    供 finalize() 在不同 σ 下重新计算模型价格，实现真正的参数扫描。
    """
    underlying_price: float   # 该 tick 的标的资产价格
    option:           Option  # 对应期权合约
    market_price:     float   # 市场（高精度 BS）报价


class BacktestCalibrationApp:
    """回测与参数校准应用。

    核心概念：
      market_engine — 代表真实市场的价格来源（vol=0.25，模拟交易所行情）
      model_engine  — 待校准的内部定价模型（初始 vol=0.15）
      校准目标：最小化 MSE(model_price, market_price)，拟合 σ → 0.25
    """

    def __init__(
        self,
        backtest_bus:   EventBus,
        main_bus:       EventBus,
        market_engine:  BlackScholesPricingEngine,  # "真实"市场价格来源
        model_engine:   BlackScholesPricingEngine,  # 待校准的内部模型
        options:        list[Option],
        calibrator:     CalibrationEngine,
        model_id:       str,
    ) -> None:
        self._backtest_bus  = backtest_bus   # 隔离回测总线（状态独立）
        self._main_bus      = main_bus       # 主系统总线（发布校准结果）
        self._market_engine = market_engine  # "市场真实"定价引擎
        self._model_engine  = model_engine   # 待校准的内部模型
        self._options       = options
        self._calibrator    = calibrator
        self._model_id      = model_id
        self._tick_count    = 0
        self._raw_obs: list[_RawObs] = []   # 原始观测缓存

    def register_handlers(self) -> None:
        """在回测专用总线上注册行情处理器（与主仿真完全隔离）。"""
        self._backtest_bus.subscribe(MarketDataEvent, self._on_market)

    def _on_market(self, event: MarketDataEvent) -> None:
        """行情事件处理器（回测重放时调用）。

        对每个期权：
          market_price = market_engine.price(opt, S).theo  （"真实"市场价）
          model_price  = model_engine.price(opt, S).theo   （模型当前预测价）
          calibrator.observe(market_price, model_price)    （记录偏差）
          raw_obs.append(...)                               （缓存原始数据）
        """
        self._tick_count += 1
        S = event.underlying_price

        for opt in self._options:
            # "市场"价格：高精度 BS（vol=0.25），模拟真实市场报价
            market_price = self._market_engine.price(opt, S).theo
            # 模型当前预测价格（vol=0.15，初始估计）
            model_price  = self._model_engine.price(opt, S).theo

            # 记录到 CalibrationEngine（用于初始 MSE 报告）
            self._calibrator.observe(market_price, model_price)
            # 缓存原始观测（供 finalize() 的 loss_fn 使用）
            self._raw_obs.append(_RawObs(S, opt, market_price))

            diff = model_price - market_price
            print(
                f"[回测|tick{self._tick_count:2d}] {opt.id}"
                f"  市场价=${market_price:.4f}"
                f"  模型价=${model_price:.4f}"
                f"  偏差=${diff:+.4f}"
            )

    def finalize(self) -> float:
        """运行参数校准，发布 ParamUpdateEvent，返回最优隐含波动率。

        流程：
          1. 构造 loss_fn(σ)：set_vol(σ) → 重算所有 raw_obs 的 MSE
          2. 调用 calibrator.solve(0.01, 1.0, loss_fn)
          3. 将最优 σ 应用到 model_engine
          4. 发布 ParamUpdateEvent 到主总线
        """
        print("\n[回测校准] ──── 开始参数优化 ────")
        print(
            f"[回测校准] 模型 ID: {self._model_id}"
            f"  原始观测数: {len(self._raw_obs)}"
            f"  初始 MSE (vol=0.15): {self._calibrator.mse():.6f}"
        )

        # ── 损失函数：给定 σ 重新计算所有观测的 MSE ─────
        # 每次调用都更新 model_engine.vol 并重新评估所有期权价格
        def loss_fn(sigma: float) -> float:
            self._model_engine.set_vol(sigma)  # 临时更新模型波动率
            if not self._raw_obs:
                return 0.0
            sum_sq = sum(
                (self._model_engine.price(obs.option, obs.underlying_price).theo
                 - obs.market_price) ** 2
                for obs in self._raw_obs
            )
            return sum_sq / len(self._raw_obs)

        # ── 黄金分割搜索：在 [0.01, 1.0] 寻找最优 σ ────
        best_vol = self._calibrator.solve(0.01, 1.0, loss_fn)

        # 将校准结果应用到模型引擎
        self._model_engine.set_vol(best_vol)
        error_pct = abs(best_vol - 0.25) / 0.25 * 100.0

        print(
            f"[回测校准] ✅ 校准完成！\n"
            f"             初始波动率: 0.1500\n"
            f"             校准波动率: {best_vol:.4f}\n"
            f"             目标波动率: 0.2500（市场真实值）\n"
            f"             校准误差:   {error_pct:.4f}%"
        )

        # ── 发布 ParamUpdateEvent 到主总线 ───────────────
        update = ParamUpdateEvent(
            model_id   = self._model_id,
            params     = {"vol": best_vol, "r": 0.05},
            updated_at = datetime.now(),
        )
        print(
            f"[回测校准] 📤 发布 ParamUpdateEvent → 模型: "
            f"{self._model_id}  参数: vol={best_vol:.4f}"
        )
        self._main_bus.publish(update)

        return best_vol

File: src/modules/seller/BacktestCalibrationApp.cpp
#include "BacktestCalibrationApp.hpp"
#include <iostream>
#include <iomanip>
#include <chrono>
#include <cmath>

// ============================================================
// BacktestCalibrationApp 实现
//
// 校准核心思路：
//   回放阶段（on_market）：收集每个 tick 的原始观测
//     raw_observations_ += { S, option, market_price }
//   校准阶段（finalize）：
//     对每个候选 σ，重新用 model_engine_ 计算所有期权价格，
//     与 market_price 比较得到 MSE 损失，黄金分割搜索最小化损失
// ============================================================

namespace omm::application {

BacktestCalibrationApp::BacktestCalibrationApp(
    std::shared_ptr<events::EventBus>              backtest_bus,
    std::shared_ptr<events::EventBus>              main_bus,
    std::shared_ptr<domain::BlackScholesPricingEngine> market_engine,
    std::shared_ptr<domain::BlackScholesPricingEngine> model_engine,
    std::vector<std::shared_ptr<domain::Option>>   options,
    std::shared_ptr<domain::CalibrationEngine>     calibrator,
    std::string                                    model_id
#ifdef BUILD_GRPC_CLIENT
    , std::shared_ptr<infrastructure::ModelServiceClient> grpc_client
#endif
)
    : backtest_bus_(std::move(backtest_bus))
    , main_bus_(std::move(main_bus))
    , market_engine_(std::move(market_engine))
    , model_engine_(std::move(model_engine))
    , options_(std::move(options))
    , calibrator_(std::move(calibrator))
    , model_id_(std::move(model_id))
    , tick_count_(0)
    , last_spot_(0.0)
#ifdef BUILD_GRPC_CLIENT
    , grpc_client_(std::move(grpc_client))
#endif
{}

void BacktestCalibrationApp::register_handlers() {
    // 在回测专用总线上订阅行情事件（与主仿真完全隔离）
    backtest_bus_->subscribe<events::MarketDataEvent>(
        [this](const events::MarketDataEvent& evt) {
            this->on_market(evt);
        }
    );
}

void BacktestCalibrationApp::on_market(const events::MarketDataEvent& event) {
    ++tick_count_;
    const double S = event.underlying_price;
    last_spot_ = S;  // 记录最新标的价格

    for (const auto& opt : options_) {
        // "市场"价格：高精度 BS（vol=0.25），模拟真实市场报价
        double market_price = market_engine_->price(*opt, S).theo;

        // 模型当前预测价（vol=0.15，初始估计）
        double model_price  = model_engine_->price(*opt, S).theo;

        // 记录到 CalibrationEngine（用于初始 MSE 报告）
        calibrator_->observe(market_price, model_price);

        // 缓存原始观测：(S, option, market_price) — 供 finalize() 的 loss_fn 使用
        raw_observations_.push_back(RawObs{S, opt, market_price});

        std::cout << "[回测|tick" << std::setw(2) << tick_count_ << "] "
                  << opt->id()
                  << "  市场价=$" << std::fixed << std::setprecision(4) << market_price
                  << "  模型价=$" << model_price
                  << "  偏差=$" << std::showpos << (model_price - market_price)
                  << std::noshowpos << "\n";
    }
}

double BacktestCalibrationApp::finalize() {
    std::cout << "\n[回测校准] ──── 开始参数优化 ────\n";
    std::cout << "[回测校准] 模型 ID: " << model_id_
              << "  原始观测数: " << raw_observations_.size()
              << "  初始 MSE (vol=0.15): "
              << std::fixed << std::setprecision(6) << calibrator_->mse() << "\n";

    // ── 损失函数：对给定 σ 重新计算所有观测的 MSE ─────────
    // 这是真正的参数化 loss(σ)，每次调用都重新评估 model_engine
    auto loss_fn = [this](double sigma) -> double {
        model_engine_->set_vol(sigma); // 临时更新模型波动率

        double sum_sq = 0.0;
        for (const auto& obs : raw_observations_) {
            // 用新 σ 重新计算模型预测价
            double model_price = model_engine_->price(*obs.option, obs.underlying_price).theo;
            double err         = model_price - obs.market_price;
            sum_sq += err * err;
        }
        return sum_sq / static_cast<double>(raw_observations_.size());
    };

    // ── 运行黄金分割搜索：在 [0.01, 1.0] 范围内寻找最优 σ ──
    double best_vol = calibrator_->solve(0.01, 1.0, loss_fn);

    // 将校准结果应用到模型引擎
    model_engine_->set_vol(best_vol);

    std::cout << "[回测校准] ✅ 校准完成！\n"
              << "             初始波动率: 0.1500\n"
              << "             校准波动率: " << std::fixed << std::setprecision(4)
              << best_vol << "\n"
              << "             目标波动率: 0.2500（市场真实值）\n"
              << "             校准误差:   "
              << std::abs(best_vol - 0.25) / 0.25 * 100.0 << "%\n";

    // ── 发布 ParamUpdateEvent 到主总线 ────────────────────
    events::ParamUpdateEvent update{
        model_id_,
        {{"vol", best_vol}, {"r", 0.05}},
        std::chrono::system_clock::now()
    };

    std::cout << "[回测校准] 📤 发布 ParamUpdateEvent → 模型: "
              << model_id_ << "  参数: vol=" << best_vol << "\n";

    main_bus_->publish(update);

#ifdef BUILD_GRPC_CLIENT
    // ── Phase 2b：粗糙 Bergomi 校准（仅当 gRPC 客户端已注入时执行）────
    if (grpc_client_) {
        std::cout << "\n[回测校准] ──── 开始 Rough Bergomi 校准（gRPC）────\n";

        // 从原始观测中提取 OptionQuote 列表（去重：用最后一次 tick 的市场价）
        std::vector<infrastructure::OptionQuote> quotes;
        for (const auto& opt : options_) {
            double mp = market_engine_->price(*opt, last_spot_).theo;
            auto now  = std::chrono::system_clock::now();
            double T  = std::chrono::duration<double>(opt->expiry() - now).count()
                        / (365.0 * 24.0 * 3600.0);
            T = std::max(T, 1e-6);
            quotes.push_back(infrastructure::OptionQuote{
                opt->strike(), T,
                opt->option_type() == domain::OptionType::Call,
                mp
            });
        }

        try {
            // n_paths=2000 — 计算轻量（"computationally light" 方案）
            auto result = grpc_client_->calibrate_rough_bergomi(
                last_spot_, quotes, 0.05, 0.0, 2000, 64, 42
            );

            std::cout << "[回测校准] ✅ Rough Bergomi 校准完成！\n"
                      << "             H   = " << result.params.at("hurst") << "\n"
                      << "             eta = " << result.params.at("eta")   << "\n"
                      << "             rho = " << result.params.at("rho")   << "\n"
                      << "             xi0 = " << result.params.at("xi0")   << "\n"
                      << "             MSE = " << result.mse
                      << "  耗时 " << result.elapsed_s << "s\n";

            // 发布粗糙参数到主总线 → ParameterStore 存储
            events::ParamUpdateEvent rough_update{
                "rough_bergomi",
                {
                    {"H",   result.params.at("hurst")},
                    {"eta", result.params.at("eta")},
                    {"rho", result.params.at("rho")},
                    {"xi0", result.params.at("xi0")}
                },
                std::chrono::system_clock::now()
            };
            std::cout << "[回测校准] 📤 发布 ParamUpdateEvent → 模型: rough_bergomi\n";
            main_bus_->publish(rough_update);

        } catch (const std::exception& e) {
            std::cout << "[回测校准] ⚠️  Rough Bergomi 校准失败（服务未启动？）: "
                      << e.what() << "\n";
        }
    }
#endif

    return best_vol;
}

} // namespace omm::application
